cmd/util/cmd/epoch-recovery/cmd/reset.go

```diff
@@ -0,0 +1,207 @@
+
+	// path to the root protocol snapshot json file
+	snapshotPath := filepath.Join(flagBootDir, bootstrap.PathRootProtocolStateSnapshot)
+
+	// check if root-protocol-snapshot.json file exists under the dir provided
+	exists, err := pathExists(snapshotPath)
+	if err != nil {
+		log.Fatal().Err(err).Str("path", snapshotPath).Msgf("could not check if root protocol-snapshot.json exists")
+	}
+	if !exists {
+		log.Fatal().Str("path", snapshotPath).Msgf("root-protocol-snapshot.json file does not exists in the `boot-dir` given")
+	}
+
+	// read root protocol-snapshot.json
+	bz, err := io.ReadFile(snapshotPath)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("could not read root snapshot file")
+	}
+
+	// unmarshal bytes to inmem protocol snapshot
+	snapshot, err := convert.BytesToInmemSnapshot(bz)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not convert array of bytes to snapshot")
+	}
+
+	// extract arguments from reset epoch tx from snapshot
+	cdcArgs := extractResetEpochArgs(snapshot)
+
+	// encode cadence arguments to JSON
+	encoded, err := jsoncdc.Encode(cdcArgs)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not encode cadence arguments")
+	}
+
+	// write JSON args to file
+	err = io.WriteFile(argsPath, encoded)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not write jsoncdc encoded arguments")
+	}
+}
+
+// extractResetEpochArgs extracts the required transaction arguments for the `resetEpoch` transaction
+func extractResetEpochArgs(snapshot *inmem.Snapshot) cadence.Array {
+
+	// get current epoch
+	epoch := snapshot.Epochs().Current()
+
+	// set payout
+	payout := uint64(0)
+
+	// read random source from epoch
+	randomSource, err := epoch.RandomSource()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get random source from epoch")
+	}
+
+	// read first view
+	firstView, err := epoch.FirstView()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get first view from epoch")
+	}
+
+	// read final view
+	finalView, err := epoch.FinalView()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get final view from epoch")
+	}
+
+	// read collector clusters and convert to strings
+	clustering, err := epoch.Clustering()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get clustering from epoch")
+	}
+
+	// read dkg public keys for all participants
+	initialIdentities, err := epoch.InitialIdentities()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get initial identities from epoch")
+	}
+
+	dkg, err := epoch.DKG()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get dkg from epoch")
+	}
+
+	dkgKeys := make([]crypto.PublicKey, 0, len(initialIdentities))
+	for _, identity := range initialIdentities {
+		keyShare, err := dkg.KeyShare(identity.NodeID)
+		if err != nil {
+			log.Fatal().Str("node_id", identity.NodeID.String()).Err(err).Msg("coiuld get key share for node id")
+		}
+		dkgKeys = append(dkgKeys, keyShare)
+	}
+
+	// TODO: read in QCs
```

Comment:
We can read the QCs using this API (https://github.com/onflow/flow-go/blob/c3c416c84087bf5140187892e4075ed85e677230/state/protocol/epoch.go#L80). The number of clusters is `len(clustering)`.

There is some trickiness here with the cluster QCs here however. Votes are gathered by the smart contract as many individual signatures, but the QC contains a single aggregated signature. When reading the protocol state, we only have access to the aggregated signature, but the form expected by the smart contract is the un-aggregated signature. The end goal here is to use aggregated signatures in both the smart contract and the protocol state to avoid this problem, but for now we have a mismatch in the representation.

However, these cluster QCs **are not used** by the smart contract when injected with `resetEpoch`. They are essentially only there as historical context, the point in which they would have been used by the contract is skipped by the epoch reset. Functionally it is OK to fudge these fields a bit. 

cc @joshuahannan Do you mind sanity checking the reasoning here, to confirm the cluster QC fields injected with `resetEpoch` are not used?

![image](https://user-images.githubusercontent.com/10557821/120541830-3ddc5b80-c39f-11eb-8794-1dd15dac8ae7.png)
[Diagram](https://swimlanes.io/d/UT15MstuU)

Reply:
correct, they are not used. They are only stored to the record of the protocol metadata